README.md
# async-unzip

Efficient, fast, asynchronous, in-memory unzip module for Node.js.

## Sample Example

	var async = require('async'),
			path = require('path'),
			ZipFile = require('async-unzip').ZipFile,
			zipFile = new ZipFile('/home/user/Name.app.dSYM.zip'),
			noMoreFiles = false;
	
	async.whilst(function () {
		return !noMoreFiles;
	}, function (cb) {
		async.waterfall([
			function (cb) {
				zipFile.getNextEntry(cb);
			},
			function (entry, cb) {
				if (entry) {
					if (entry.isFile) {
						var match = entry.filename.match(/^[^\/]+\.dSYM\/Contents\/Resources\/DWARF\/(.+)/);
	
						if (match) {
							console.log(match);
							console.log(entry);
	
							async.waterfall([
								function (cb) {
									entry.getData(cb);
								},
								function (data, cb) {
									// `data` is a binary data of the entry.
									console.log(data.length);
	
									cb();
								}
							], cb);
						}
					}
				} else {
					noMoreFiles = true;
				}
	
				cb();
			}
		], cb);
	}, function () {
		// DO NOT FORGET to call `close()` to release the open file descriptor,
		// otherwise, you will quickly run out of file descriptors.
		zipFile.close();
		console.log(arguments);
	});


CHANGELOG.md
# async-unzip Change Log

## Version 0.1.3 (Nov 6, 2014)

- FIX: the compressed data is not necessarily right after the local header. Read separately.

## Version 0.1.2 (Aug 28, 2014)

- FIX: correctly determine if the ZIP entry is a file or directory.

## Version 0.1.1 (Aug 28, 2014)

- FIX: release open file descriptors.

## Version 0.1.0 (Jun 19, 2014)

- NEW: initial release.


index.js
var async = require('async'),
		CentralDirectory = require('./CentralDirectoryFileHeader');

var ZipFile = module.exports.ZipFile = function(/*String*/filePath) {
	if (!(this instanceof ZipFile)) {
		return new ZipFile(filePath);
	}

	this.filePath = filePath;
};

ZipFile.prototype.getCentralDirectory = function (cb) {
	if (this._centralDirectory) {
		cb(null, this._centralDirectory);
	} else {
		var self = this;

		async.waterfall([
			function (cb) {
				CentralDirectory.initCentralDirectory(self, cb);
			},
			function (centralDirectory, cb) {
				self._centralDirectory = centralDirectory;

				cb(null, centralDirectory);
			}
		], cb);
	}
};

ZipFile.prototype.getNextEntry = function (cb) {
	var self = this;

	async.waterfall([
		function (cb) {
			self.getCentralDirectory(cb);
		},
		function (centralDirectory, cb) {
			centralDirectory.getNextEntry(cb);
		}
	], cb);
};

ZipFile.prototype.close = function (cb) {
	!cb && (cb = function () {});

	if (this._centralDirectory) {
		this._centralDirectory.close(cb);
	} else {
		cb();
	}
};


ZipEntry.js
var async = require('async'),
		fs = require('fs'),
		LocalFileHeader = require('./LocalFileHeader');

var ZipEntry;

var CONSTANTS = {
	CD_HEADER_SIZE                         : 46,
	CD_SIGNATURE                           : 0x02014b50,
	OFFSETS: {
		CD_SIGNATURE                         : 0,
		VERSION_MADE_BY                      : 4,
		MIN_VERSION_NEEDED_TO_EXTRACT        : 6,
		GENERAL_PURPOSE_BIT_FLAG             : 8, // encrypt, decrypt flags
		COMPRESSION_METHOD                   : 10,
		LAST_MODIFIED_TIME                   : 12,
		LAST_MODIFIED_DATE                   : 14,
		CRC32                                : 16,
		COMPRESSED_SIZE                      : 20, // compressed size
		UNCOMPRESSED_SIZE                    : 24, // uncompressed size
		FILENAME_LENGTH                      : 28, // filename length
		EXTRA_FIELD_LENGTH                   : 30, // extra field length
		FILE_COMMENT_LENGTH                  : 32, // file comment length
		DISK_NUMBER_WHERE_FILE_STARTS        : 34, // volume number start
		INTERNAL_FILE_ATTRIBUTES             : 36, // internal file attributes
		EXTERNAL_FILE_ATTRIBUTES             : 38, // external file attributes (host system dependent)
		RELATIVE_OFFSET_OF_LOCAL_FILE_HEADER : 42 // LOC header offset
	},
	ERRORS: {
		INVALID_DATA_OFFSET: 'Invalid file data offset.'
	},
	COMPRESSION_METHODS: {
		DEFLATE: 8
	}
};

ZipEntry = module.exports = function (centralDirectory) {
	var buffer = centralDirectory._buffer,
			offset = centralDirectory._offset;

	this._fd = centralDirectory._fd;
	this.bitFlag = buffer.readUInt16LE(offset + CONSTANTS.OFFSETS.GENERAL_PURPOSE_BIT_FLAG);

	// When compression method is 8, `Deflate` compression is used.
	this.compressionMethod = buffer.readUInt16LE(offset + CONSTANTS.OFFSETS.COMPRESSION_METHOD);

	this.compressedSize = buffer.readUInt32LE(offset + CONSTANTS.OFFSETS.COMPRESSED_SIZE);
	this.uncompressedSize = buffer.readUInt32LE(offset + CONSTANTS.OFFSETS.UNCOMPRESSED_SIZE);
	this.filenameLength = buffer.readUInt16LE(offset + CONSTANTS.OFFSETS.FILENAME_LENGTH);
	this.extraFieldLength = buffer.readUInt16LE(offset + CONSTANTS.OFFSETS.EXTRA_FIELD_LENGTH);
	this.fileCommentLength = buffer.readUInt16LE(offset + CONSTANTS.OFFSETS.FILE_COMMENT_LENGTH);
	this.localFileHeaderOffset = buffer.readUInt32LE(offset + CONSTANTS.OFFSETS.RELATIVE_OFFSET_OF_LOCAL_FILE_HEADER);

	offset += CONSTANTS.CD_HEADER_SIZE;

	this.filename = buffer.toString('utf8', offset, offset += this.filenameLength);

	this.entrySize = CONSTANTS.CD_HEADER_SIZE + this.filenameLength + this.extraFieldLength + this.fileCommentLength;

	// According Open JDK source code at
	// http://grepcode.com/file/repository.grepcode.com/java/root/jdk/openjdk/6-b14/java/util/zip/ZipEntry.java#ZipEntry.isDirectory%28%29
	// this is the way to determine if a ZIP entry a file or directory.
	this.isFile = this.filename.slice(-1) != '/';
};

ZipEntry.prototype.getData = function (size, cb) {
	var localFileHeader = new LocalFileHeader(this);

	localFileHeader.getData(size, cb);
};


CentralDirectoryFileHeader.js
/*
 * Central directory file header
 * Refer to http://en.wikipedia.org/wiki/Zip_(file_format).
 Offset	Bytes	Description
  0   	4   	Central directory file header signature = 0x02014b50
  4   	2   	Version made by
  6   	2   	Version needed to extract (minimum)
  8   	2   	General purpose bit flag
 10   	2   	Compression method
 12   	2   	File last modification time
 14   	2   	File last modification date
 16   	4   	CRC-32
 20   	4   	Compressed size
 24   	4   	Uncompressed size
 28   	2   	File name length (n)
 30   	2   	Extra field length (m)
 32   	2   	File comment length (k)
 34   	2   	Disk number where file starts
 36   	2   	Internal file attributes
 38   	4   	External file attributes
 42   	4   	Relative offset of local file header. This is the number of bytes between the start of the first disk on which the file occurs, and the start of the local file header. This allows software reading the central directory to locate the position of the file inside the .ZIP file.
 46   	n   	File name
 46+n   m   	Extra field
 46+n+m	k   	File comment
* */
var async = require('async'),
		fs = require('fs'),
		EndOfCentralDirectory = require('./EndOfCentralDirectoryRecord'),
		ZipEntry = require('./ZipEntry');

var CONSTANTS = {
	CD_HEADER_SIZE                         : 46,
	CD_SIGNATURE                           : 0x02014b50,
	OFFSETS: {
		CD_SIGNATURE                         : 0,
		VERSION_MADE_BY                      : 4,
		MIN_VERSION_NEEDED_TO_EXTRACT        : 6,
		GENERAL_PURPOSE_BIT_FLAG             : 8, // encrypt, decrypt flags
		COMPRESSION_METHOD                   : 10,
		LAST_MODIFIED_TIME                   : 12,
		LAST_MODIFIED_DATE                   : 14,
		CRC32                                : 16,
		COMPRESSED_SIZE                      : 20, // compressed size
		UNCOMPRESSED_SIZE                    : 24, // uncompressed size
		FILENAME_LENGTH                      : 28, // filename length
		EXTRA_FIELD_LENGTH                   : 30, // extra field length
		FILE_COMMENT_LENGTH                  : 32, // file comment length
		DISK_NUMBER_WHERE_FILE_STARTS        : 34, // volume number start
		INTERNAL_FILE_ATTRIBUTES             : 36, // internal file attributes
		EXTERNAL_FILE_ATTRIBUTES             : 38, // external file attributes (host system dependent)
		RELATIVE_OFFSET_OF_LOCAL_FILE_HEADER : 42 // LOC header offset
	},
	ERRORS: {
		INVALID_CD: 'Invalid Central Directory Record.',
		INCOMPLETE_CD: 'Incomplete Central Directory Record.'
	}
};

module.exports.initCentralDirectory = function (zipFile, cb) {
	var self = new CentralDirectory(),
			eocd = new EndOfCentralDirectory(zipFile);

	async.waterfall([
		function (cb) {
			eocd.getCentralDirectoryMeta(cb);
		},
		function (meta, cb) {
			self._sizeInBytes = meta.sizeInBytes;
			self._numberOfEntries = meta.numberOfEntries;
			self._fd = meta.fd;

			if (meta.offset + meta.sizeInBytes > eocd._stat.size) {
				cb(new Error(CONSTANTS.ERRORS.INCOMPLETE_CD));
			} else {
				fs.read(eocd._fd, new Buffer(meta.sizeInBytes), 0, meta.sizeInBytes, meta.offset, cb);
			}
		},
		function (bytesRead, buffer, cb) {
			if (bytesRead == self._sizeInBytes) {
				self._buffer = buffer;

				cb(null, self);
			} else {
				cb(new Error(CONSTANTS.ERRORS.INVALID_CD));
			}
		}
	], cb);
};

var CentralDirectory = function () {
	this._offset = 0;
};

CentralDirectory.prototype.getNextEntry = function (cb) {
	var entry;

	if (this._offset < this._sizeInBytes) {
		entry = new ZipEntry(this);
		this._offset += entry.entrySize;
	}

	cb(null, entry);
};

CentralDirectory.prototype.close = function (cb) {
	fs.close(this._fd, cb);
};


EndOfCentralDirectoryRecord.js
/*
 * End of central directory record (EOCD).
 * Refer to http://en.wikipedia.org/wiki/Zip_(file_format).
 Offset	Bytes	Description
  0	    4	    End of central directory signature = 0x06054b50
  4	    2	    Number of this disk
  6	    2	    Disk where central directory starts
  8	    2	    Number of central directory records on this disk
 10	    2	    Total number of central directory records
 12	    4	    Size of central directory (bytes)
 16	    4	    Offset of start of central directory, relative to start of archive
 20	    2	    Comment length (n)
 22	    n	    Comment
* */
var async = require('async'),
		fs = require('fs');

var CONSTANTS = {
	EOCD_HEADER_SIZE                        : 22, // END header size
	EOCD_SIGNATURE                          : 0x06054b50, // "PK\005\006"
	OFFSETS: {
		EOCD_SIGNATURE                        : 0,
		NUMBER_OF_THIS_DISK                   : 4,
		DISK_WHERE_CENTRAL_DIRECTORY_STARTS   : 6,
		NUMBER_OF_ENTRIES_ON_THIS_DISK        : 8,
		TOTAL_NUMBER_OF_ENTRIES_ON_THIS_DISK  : 10,
		CD_SIZE_IN_BYTES                      : 12,
		CD_START                              : 16,
		COMMENT_LENGTH                        : 20,
		COMMENT_START                         : 22
	},
	ERRORS: {
		INVALID_EOCD: 'Invalid end of central directory record.'
	}
};

var EOCD = module.exports = function (zipFile) {
	this._zipFile = zipFile;
};

EOCD.prototype.readHeader = function (cb) {
	if (this._buffer) {
		cb(null, this._buffer);
	} else {
		var self = this;

		async.waterfall([
			function (cb) {
				fs.open(self._zipFile.filePath, 'r', cb);
			},
			function (fd, cb) {
				self._fd = fd;

				fs.fstat(fd, cb);
			},
			function (stat, cb) {
				self._stat = stat;

				fs.read(self._fd, new Buffer(CONSTANTS.EOCD_HEADER_SIZE), 0, CONSTANTS.EOCD_HEADER_SIZE, stat.size - CONSTANTS.EOCD_HEADER_SIZE, cb);
			},
			function (bytesRead, buffer, cb) {
				if (bytesRead == CONSTANTS.EOCD_HEADER_SIZE) {
					self._buffer = buffer;

					cb(null, buffer);
				} else {
					cb(new Error(CONSTANTS.ERRORS.INVALID_EOCD));
				}
			}
		], cb);
	}
};

EOCD.prototype.getCentralDirectoryMeta = function (cb) {
	var self = this;

	async.waterfall([
		function (cb) {
			self.readHeader(cb);
		},
		function (buffer, cb) {
			cb(null, {
				fd: self._fd,
				numberOfEntries: buffer.readUInt16LE(CONSTANTS.OFFSETS.NUMBER_OF_ENTRIES_ON_THIS_DISK),
				offset: buffer.readUInt32LE(CONSTANTS.OFFSETS.CD_START),
				sizeInBytes: buffer.readUInt32LE(CONSTANTS.OFFSETS.CD_SIZE_IN_BYTES)
			});
		}
	], cb);
};


LocalFileHeader.js
/*
 *  Local file header
 * Refer to http://en.wikipedia.org/wiki/Zip_(file_format).
 Offset	Bytes	Description
  0	    4		  Local file header signature = 0x04034b50 (read as a little-endian number)
  4	  	2		  Version needed to extract (minimum)
  6		  2		  General purpose bit flag
  8		  2		  Compression method
 10		  2	  	File last modification time
 12		  2	  	File last modification date
 14		  4	  	CRC-32
 18		  4	  	Compressed size
 22		  4	  	Uncompressed size
 26		  2	  	File name length (n)
 28		  2	  	Extra field length (m)
 30		  n	  	File name
 30+n	  m	  	Extra field
 * */
var async = require('async'),
		fs = require('fs'),
		zlib = require('zlib');

var LocalFileHeader;

var CONSTANTS = {
	LOCAL_FILE_HEADER_SIZE          : 30,
	LOCAL_FILE_HEADER_SIGNATURE     : 0x04034b50,
	OFFSETS: {
		LOCAL_FILE_HEADER_SIGNATURE   : 0,
		MIN_VERSION_NEEDED_TO_EXTRACT : 4,
		GENERAL_PURPOSE_BIT_FLAG      : 6, // encrypt, decrypt flags
		COMPRESSION_METHOD            : 8,
		LAST_MODIFIED_TIME            : 10,
		LAST_MODIFIED_DATE            : 12,
		CRC32                         : 14,
		COMPRESSED_SIZE               : 18,
		UNCOMPRESSED_SIZE             : 22,
		FILENAME_LENGTH               : 26,
		EXTRA_FIELD_LENGTH            : 28,
		FILENAME                      : 30
	},
	ERRORS: {
		INVALID_COMPRESSION_METHOD: 'Invalid or unsupported file compression method.',
		INVALID_DATA_OFFSET: 'Invalid file data offset.',
		INVALID_LOCAL_HEADER_OFFSET: 'Invalid local header offset.',
		INVALID_FILE_HEADER_SIGNATURE: 'Invalid file header signature.'
	},
	COMPRESSION_METHODS: {
		STORED: 0,
		DEFLATED: 8
	},
	// In Zlib, this is the default amount for chunk size:
	// `#define CHUNK 16384`
	MIN_BYTES_TO_READ: 16*1024
};

LocalFileHeader = module.exports = function (entry) {
	if (!this instanceof LocalFileHeader) {
		return new LocalFileHeader(entry);
	}

	this._entry = entry;
};

LocalFileHeader.prototype.readLocalFileHeader = function (buffer, cb) {
	this.signature = buffer.readUInt32LE(CONSTANTS.OFFSETS.LOCAL_FILE_HEADER_SIGNATURE);

	if (this.signature !== CONSTANTS.LOCAL_FILE_HEADER_SIGNATURE) {
		cb(new Error(CONSTANTS.ERRORS.INVALID_FILE_HEADER_SIGNATURE));
	} else {
		this.filenameLength = buffer.readUInt16LE(CONSTANTS.OFFSETS.FILENAME_LENGTH);
		this.extraFieldLength = buffer.readUInt16LE(CONSTANTS.OFFSETS.EXTRA_FIELD_LENGTH);

		this.fileDataOffset = CONSTANTS.LOCAL_FILE_HEADER_SIZE + this.filenameLength + this.extraFieldLength;

		cb();
	}
};

LocalFileHeader.prototype.getData = function (size, cb) {
	var compressedReadSize,
			entry = this._entry,
			self = this;

	if (typeof size === 'function') {
		cb = size;
		size = entry.uncompressedSize;
		compressedReadSize = entry.compressedSize;
	} else {
		if (size > entry.compressedSize) {
			compressedReadSize = entry.compressedSize;

			size > entry.uncompressedSize && (size = entry.uncompressedSize);
		} else if (size < CONSTANTS.MIN_BYTES_TO_READ) {
			compressedReadSize = CONSTANTS.MIN_BYTES_TO_READ;

			size < 0 && (size = 0);
		} else {
			// Read twice more compressed data from the file system to avoid
			// [Error: invalid distance too far back] errno: -3, code: 'Z_DATA_ERROR' }
			// when inflating an incomplete compressed data chunks.
			compressedReadSize = size * 2;
		}
	}

	async.waterfall([
		function (cb) {
			// We first need to read the header to determine where the
			// file data starts, i.e. the file data offset.
			// The thing is it doesn't always start right after the
			// header data. There may be some other info in between.
			fs.read(entry._fd, new Buffer(CONSTANTS.LOCAL_FILE_HEADER_SIZE), 0, CONSTANTS.LOCAL_FILE_HEADER_SIZE, entry.localFileHeaderOffset, cb);
		},
		function (bytesRead, headBuffer, cb) {
			if (bytesRead == CONSTANTS.LOCAL_FILE_HEADER_SIZE) {
				self.readLocalFileHeader(headBuffer, cb);
			} else {
				cb(new Error(CONSTANTS.ERRORS.INVALID_LOCAL_HEADER_OFFSET));
			}
		},
		function (cb) {
			// Now read the actual data which can be either the compressed data
			// or already uncompressed, i.e. stored data.
			fs.read(entry._fd, new Buffer(compressedReadSize), 0, compressedReadSize, entry.localFileHeaderOffset + self.fileDataOffset, cb);
		},
		function (bytesRead, buffer, cb) {
			if (bytesRead == compressedReadSize) {
				if (entry.compressionMethod == CONSTANTS.COMPRESSION_METHODS.STORED) {
					cb(null, buffer);
				} else if (entry.compressionMethod == CONSTANTS.COMPRESSION_METHODS.DEFLATED) {
					var buff = new Buffer(size),
							inflator = zlib.createInflateRaw(),
							written = 0;

					inflator.once('end', onEnd);

					inflator.on('error', function (err) {
						inflator.removeListener('end', onEnd);
						inflator.removeListener('readable', processChunk);

						cb(err);
					});

					function onEnd() {
						cb(null, buff);
					}

					function processChunk() {
						var chunk;

						while (written < size && (chunk = inflator.read())) {
							var len = size - written;

							chunk.length < len && (len = chunk.length);
							chunk.copy(buff, written, 0, len);

							written += len;
						}

						if (written < size) {
							inflator.once('readable', processChunk);
						}
					}

					// Pass in the compressed data.
					inflator.end(buffer);

					processChunk();
				} else {
					cb(new Error(CONSTANTS.ERRORS.INVALID_COMPRESSION_METHOD));
				}
			} else {
				cb(new Error(CONSTANTS.ERRORS.INVALID_DATA_OFFSET));
			}
		}
	], cb);
};
